### PR TITLE
test: e2e: framework: add a simple artifact saver

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Deferring the new framework.Artifact() will write a YAML file with the
object that you're asking to be saved. The file will be placed under
directories as necessary for the logical cluster and namespace it's in,
and named by kind.group_name.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

fixes https://github.com/kcp-dev/kcp/issues/271

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Output looks like:

```
$ tree /tmp/e2e/TestClusterController/update_a_synced_object,_expect_status_to_sync_to_source/1720278361/artifacts/
/tmp/e2e/TestClusterController/update_a_synced_object,_expect_status_to_sync_to_source/1720278361/artifacts/
└── kcp
    ├── sink
    │   ├── 192-168-0-243-43617---admin
    │   │   └── cluster-controller-test
    │   │       └── cowboys.wildwest.dev_timothy.yaml
    │   └── kcp.log
    └── source
        ├── 192-168-0-243-42007---admin
        │   ├── cluster-controller-test
        │   │   └── cowboys.wildwest.dev_timothy.yaml
        │   └── clusters.cluster.example.dev_us-east1.yaml
        └── kcp.log
```

File: test/e2e/framework/util.go
Comment (stevekuznetsov) [inline test/e2e/framework/util.go:87]: Fundamentally I wanted to ask the user for the smallest amount of input to this, which means we end up doing some "extra" things like using the RESTMapper to find a nice name for the artifact and messing with the Scheme so we ensure that we're serializing something sensible to disk that can be `kubectl apply`-ed to another `kcp` later.

Comment (stevekuznetsov) [reply test/e2e/framework/util.go:87]: After a bit more thought, this might make more sense to hang off of the `RunningServer` since that knows how to e.g. get the discovery client for the thing we're talking to, and knows its own name, which are both things that the user should not bother needing to pass in. I might do another refactor pass

File: Makefile
Comment (sttts) [inline Makefile:80]: mixed with other PR/commit

Comment (stevekuznetsov) [reply Makefile:80]: Yeah, I broke it out from this one - can rebase once it lands

**Review by ncdc — APPROVED** [review]
LGTM, ready to merge if you are

**Pull request merged**
